src/main/java/spoon/support/reflect/declaration/CtNamedElementImpl.java

```diff
@@ -34,21 +36,35 @@ public String getSimpleName() {
 
 	@Override
 	public <T extends CtNamedElement> T setSimpleName(String simpleName) {
+		String oldName = this.simpleName;
+
 		Factory factory = getFactory();
 		if (factory == null) {
 			this.simpleName = simpleName;
+			updateParentPackage(oldName);
```

Comment:
I don't like the idea of updating the parent package when setting the simple name, but I also can't really see a better solution. I've pondered it for a good 20 minutes now, and I can't figure something out that doesn't end up violating the contract of the set (or keys of the map, same thing) referencing the type/package. Mutability is terrible, isn't it? With the current implementation in Spoon, you can for example get two different types with the same name in a single package, which is all kinds of crazy.

Anyway, this is not the concern of `CtNamedElement`, but rather should be implemented in `CtTypeImpl` and `CtPackageImpl`. When you need to check the type of `this`, you're probably up to no good :)

Finally, there is the concern of what should happen if you change the name of a type/package such that it collides with another one in the same package. The current behavior appears to just clobber in case of a collision. This has been discussed before, and we found that an exception would be best: #3837

Reply:
And this vomits errors all over the test suite because apparently this is done in multiple places. Another question is how this should interact with `setIgnoreDuplicateDeclarations`:
```java
	/**
	 * Set Spoon to ignore duplicate type declarations in a project.
	 *
	 * Setting this option to {@code true} causes Spoon to attempt to build a model even when
	 * the same qualified name appears for multiple types. This mode of operation makes Spoon less
	 * stable as duplicated types do not make sense in Java, and causes strange behavior in the
	 * underlying JDT compiler. Type resolution can become unpredictable as the order in which types
	 * are parsed becomes a determining factor in which duplicated type actually makes it into the
	 * model.
     *
	 * @param ignoreDuplicateDeclarations (default false)  set to true to allow spoon to create a model of a project that
	 *                                 contains multiple times the same class
	 */
	void setIgnoreDuplicateDeclarations(boolean ignoreDuplicateDeclarations);
```

Reply:
Just to be clear, this changes behavior. Previously, if you changed the simple name of a type such that it collided with another in the same package, they would _both_ be in the set (and thus violating the contract of that set).

With your implementation here, one overwrites the other.

Anyway, since there turned out to be problems with the stricter enforcement, let's leave that for a separate PR. But, do document the current "overwriting" behavior with a test somehow. Show that if you rename such that there is a clash, the renamed element is the one that remains in the package. Understand what I mean?

Reply:
I added a test: https://github.com/INRIA/spoon/pull/4076/commits/b2a2ca77b60f1e9bb72d3a873b522c57c5ae4afa

Reply:
I was thinking of high-level tests on CtType.setSimpleName() and CtPackage.setSimpleName(). I.e. given a package with types A and B, calling B.setSImpleName("A") should result in the "original" A being removed from the package. Which, if I'm not misreading your code, is what happens with your implementation?

Reply:
It doesn't change the simple name, it only validates the add method as that contained the logic the other tests depend on. I will add some tests that explicitly check the rename as well, that was another aspect of this implementation and not quite what we talked about earlier in this thread.

Reply:
To be super clear about what I mean, try this code snippet on master and with this PR:

```java
Launcher launcher = new Launcher();
CtModel model = launcher.buildModel();
Factory factory = launcher.getFactory();

CtType<?> bestClass = factory.createClass("BestClass");
CtType<?> otherCLass = factory.createClass("OtherClass");

System.out.println(model.getRootPackage().getTypes());

bestClass.setSimpleName(otherClass.getSimpleName());

System.out.println(model.getRootPackage().getTypes());
```